### PR TITLE
[cmd/telemetrygen] Use exporter per worker for better metrics throughput

### PR DESCRIPTION
**Pull request by marcelbirkner**
**Description:**
Adding a feature - Use exporter per worker for better metrics throughput

Initially when adding more workers in the telemetrygen config when running "metrics" it did not increase the metrics throughput since all workers used the same exporter. 

By creating one exporter per worker we can now increase the number of metrics being send to the backend. 

**Link to tracking Issue:** https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/26709

**Testing:** 
- Units tests pass
- Ran local load tests with different configurations

## Before code change

Generate metrics:

```
telemetrygen metrics \
    --metric-type Sum \
    --duration "60s" \
    --rate "0" \
    --workers "10" \
    --otlp-http=false \
    --otlp-endpoint <HOSTNAME> \
    --otlp-attributes "service.name"=\"telemetrygen\"
```

Output:
```
metrics generated	{"worker": 8, "metrics": 139}
metrics generated	{"worker": 0, "metrics": 139}
metrics generated	{"worker": 9, "metrics": 141}
metrics generated	{"worker": 4, "metrics": 140}
metrics generated	{"worker": 2, "metrics": 140}
metrics generated	{"worker": 3, "metrics": 140}
metrics generated	{"worker": 7, "metrics": 140}
metrics generated	{"worker": 5, "metrics": 140}
metrics generated	{"worker": 1, "metrics": 140}
metrics generated	{"worker": 6, "metrics": 140}
```

## After code change

```
telemetrygen metrics \
    --metric-type Sum \
    --duration "60s" \
    --rate "0" \
    --workers "10" \
    --otlp-http=false \
    --otlp-endpoint <HOSTNAME> \
    --otlp-attributes "service.name"=\"telemetrygen\"
```

Output:

```
metrics generated	{"worker": 6, "metrics": 1292}
metrics generated	{"worker": 3, "metrics": 1277}
metrics generated	{"worker": 5, "metrics": 1272}
metrics generated	{"worker": 8, "metrics": 1251}
metrics generated	{"worker": 9, "metrics": 1241}
metrics generated	{"worker": 4, "metrics": 1227}
metrics generated	{"worker": 0, "metrics": 1212}
metrics generated	{"worker": 2, "metrics": 1201}
metrics generated	{"worker": 1, "metrics": 1333}
metrics generated	{"worker": 7, "metrics": 1363}
```

By adding more workers you can now export more metrics and use `telemetrygen` better for load testing use cases.

With the code change I can now utilize my CPU better for load tests. When adding 200 workers to the above config the CPU usage can go above 80%. Before that CPU usage would be around 1% with 200 workers.

image

### CODE REVIEW

**Comment by marcelbirkner** [comment]
GH action error is unrelated to code change. Looks like intermittent error with codecov.io

```
Error uploading to [https://codecov.io:](https://codecov.io/) Error: There was an error fetching the storage URL during POST: 404 - {'detail': ErrorDetail(string='Unable to locate build via Github Actions API. Please upload with the Codecov repository upload token to resolve issue.', code='not_found')}
```

File: cmd/telemetrygen/internal/metrics/worker_test.go [hunk 1]
```diff
@@ -92,11 +99,16 @@ func TestUnthrottled(t *testing.T) {
 		},
 		MetricType: metricTypeSum,
 	}
-	exp := &mockExporter{}
+
+	m := &mockExporter{}
+	expFunc := func() (sdkmetric.Exporter, error) {
+
```
Comment (codeboten) [inline cmd/telemetrygen/internal/metrics/worker_test.go:105]: nit
```suggestion
```

File: cmd/telemetrygen/internal/metrics/worker_test.go [hunk 2]
```diff
@@ -72,16 +75,20 @@ func TestRateOfMetrics(t *testing.T) {
...
-	require.NoError(t, Run(cfg, exp, zap.NewNop()))
+	require.NoError(t, Run(cfg, expFunc, zap.NewNop()))
 
 	// verify
 	// the minimum acceptable number of metrics for the rate of 10/sec for half a second
-	assert.True(t, len(exp.rms) >= 6, "there should have been more than 6 metrics, had %d", len(exp.rms))
-	// the maximum acceptable number of metrics for the rate of 10/sec for half a second
-	assert.True(t, len(exp.rms) <= 20, "there should have been less than 20 metrics, had %d", len(exp.rms))
+	assert.True(t, len(m.rms) >= 6, "there should have been more than 6 metrics, had %d", len(m.rms))
+	// // the maximum acceptable number of metrics for the rate of 10/sec for half a second
```
Comment (codeboten) [inline cmd/telemetrygen/internal/metrics/worker_test.go:90]:
```suggestion
	// the maximum acceptable number of metrics for the rate of 10/sec for half a second
```

**Review by codeboten — COMMENTED** [review]
Thanks for the change, just a couple of minor suggestions. Does it make sense to do the same for traces and logs?

**Comment by marcelbirkner** [comment]
> Thanks for the change, just a couple of minor suggestions. Does it make sense to do the same for traces and logs?

I will have a look at traces and logs beginning next week and test if a similar change will make sense.

**Comment by marcelbirkner** [comment]
# Results for Logs

Regarding making the same changes for logs. 

TLDR: 
- Throughput is not higher when using 1 exporter per worker
- CPU usage increases but no better throuput

# Test results

I made a similar code change like with metrics where I am using 1 exporter per worker. The following shows my test results. I was testing **telemetrygen** with 1/10/50/200/1000 workers when sending logs.

I collected average CPU usage for telemetrygen as well as the number of logs each worker exported.

- BEFORE: uses the current version of telemetrygen for exporting logs
- AFTER: uses the modified version with 1 exporter per worker

## 1 Worker

BEFORE
```
❯ ./telemetrygen  logs --duration 60s --otlp-insecure --workers 1 --otlp-endpoint 100.124.42.153:4317
1.5% CPU

logs generated	{"worker": 0, "logs": 2039}
```

AFTER
```
❯ ./telemetrygen_darwin_arm64  logs --duration 60s --otlp-insecure --workers 1 --otlp-endpoint 100.124.42.153:4317
1.5% CPU

logs generated	{"worker": 0, "logs": 2059}
```

## 10 Workers

BEFORE
```
❯ ./telemetrygen  logs --duration 60s --otlp-insecure --workers 10 --otlp-endpoint 100.124.42.153:4317
3% CPU

logs generated	{"worker": 8, "logs": 2007}
logs generated	{"worker": 9, "logs": 2008}
logs generated	{"worker": 1, "logs": 2005}
logs generated	{"worker": 0, "logs": 2005}
logs generated	{"worker": 4, "logs": 2004}
logs generated	{"worker": 3, "logs": 2008}
logs generated	{"worker": 5, "logs": 2004}
logs generated	{"worker": 6, "logs": 2007}
logs generated	{"worker": 2, "logs": 2005}
logs generated	{"worker": 7, "logs": 2005}
``

AFTER
```
./telemetrygen_darwin_arm64  logs --duration 60s --otlp-insecure --workers 10 --otlp-endpoint 100.124.42.153:4317
6% CPU

logs generated	{"worker": 1, "logs": 1849}
logs generated	{"worker": 6, "logs": 1857}
logs generated	{"worker": 3, "logs": 1850}
logs generated	{"worker": 9, "logs": 1852}
logs generated	{"worker": 0, "logs": 1854}
logs generated	{"worker": 8, "logs": 1848}
logs generated	{"worker": 7, "logs": 1856}
logs generated	{"worker": 5, "logs": 1856}
logs generated	{"worker": 4, "logs": 1856}
logs generated	{"worker": 2, "logs": 1851}
```

## 50 Workers

BEFORE
```
❯ ./telemetrygen  logs --duration 60s --otlp-insecure --workers 50 --otlp-endpoint 100.124.42.153:4317
8% CPU

logs generated	{"worker": 17, "logs": 1842}
logs generated	{"worker": 25, "logs": 1829}
logs generated	{"worker": 42, "logs": 1840}
logs generated	{"worker": 20, "logs": 1832}
logs generated	{"worker": 22, "logs": 1839}
logs generated	{"worker": 8, "logs": 1841}
logs generated	{"worker": 44, "logs": 1836}
logs generated	{"worker": 32, "logs": 1832}
logs generated	{"worker": 34, "logs": 1834}
logs generated	{"worker": 41, "logs": 1835}
...
```

AFTER
```
./telemetrygen_darwin_arm64  logs --duration 60s --otlp-insecure --workers 50 --otlp-endpoint 100.124.42.153:4317
16% CPU

logs generated	{"worker": 27, "logs": 1599}
logs generated	{"worker": 40, "logs": 1600}
logs generated	{"worker": 3, "logs": 1603}
logs generated	{"worker": 30, "logs": 1593}
logs generated	{"worker": 43, "logs": 1614}
logs generated	{"worker": 38, "logs": 1598}
logs generated	{"worker": 19, "logs": 1606}
logs generated	{"worker": 0, "logs": 1601}
logs generated	{"worker": 7, "logs": 1595}
logs generated	{"worker": 36, "logs": 1604}
logs generated	{"worker": 49, "logs": 1600}
logs generated	{"worker": 26, "logs": 1616}
...
```

## 200 Workers

BEFORE
```
❯ ./telemetrygen  logs --duration 60s --otlp-insecure --workers 50 --otlp-endpoint 100.124.42.153:4317
31% CPU

logs generated	{"worker": 113, "logs": 1708}
logs generated	{"worker": 32, "logs": 1726}
logs generated	{"worker": 149, "logs": 1714}
logs generated	{"worker": 183, "logs": 1725}
logs generated	{"worker": 41, "logs": 1713}
logs generated	{"worker": 79, "logs": 1717}
logs generated	{"worker": 65, "logs": 1715}
logs generated	{"worker": 81, "logs": 1715}
logs generated	{"worker": 127, "logs": 1714}
```


AFTER
```
./telemetrygen_darwin_arm64  logs --duration 60s --otlp-insecure --workers 200 --otlp-endpoint 100.124.42.153:4317
40% CPU

logs generated	{"worker": 26, "logs": 1162}
logs generated	{"worker": 15, "logs": 1211}
logs generated	{"worker": 127, "logs": 1215}
logs generated	{"worker": 97, "logs": 1156}
logs generated	{"worker": 83, "logs": 1152}
logs generated	{"worker": 155, "logs": 1127}
logs generated	{"worker": 140, "logs": 1129}
logs generated	{"worker": 178, "logs": 1130}
logs generated	{"worker": 147, "logs": 1140}
logs generated	{"worker": 16, "logs": 1145}
logs generated	{"worker": 20, "logs": 1218}
logs generated	{"worker": 94, "logs": 1175}
logs generated	{"worker": 24, "logs": 1120}
logs generated	{"worker": 145, "logs": 1187}
logs generated	{"worker": 58, "logs": 1171}
...
```

## 1000 Workers

BEFORE
```
❯ ./telemetrygen  logs --duration 60s --otlp-insecure --workers 1000 --otlp-endpoint 100.124.42.153:4317
99% CPU

logs generated	{"worker": 865, "logs": 1213}
logs generated	{"worker": 684, "logs": 1212}
logs generated	{"worker": 840, "logs": 1215}
logs generated	{"worker": 60, "logs": 1207}
logs generated	{"worker": 678, "logs": 1214}
logs generated	{"worker": 229, "logs": 1215}
```

AFTER
```
./telemetrygen_darwin_arm64  logs --duration 60s --otlp-insecure --workers 1000 --otlp-endpoint 100.124.42.153:4317
70% CPU

logs generated	{"worker": 116, "logs": 278}
logs generated	{"worker": 587, "logs": 342}
logs generated	{"worker": 149, "logs": 338}
logs generated	{"worker": 184, "logs": 224}
logs generated	{"worker": 278, "logs": 325}
logs generated	{"worker": 588, "logs": 252}
logs generated	{"worker": 111, "logs": 299}
logs generated	{"worker": 106, "logs": 171}
...
```

**Review by mx-psi — APPROVED** [review]

**Review by codeboten — APPROVED** [review]
Thanks @marcelbirkner!
